File: tools/dictionary/repair_desk.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Loopback-only review desk for queuing dictionary correction candidates.

This server deliberately keeps candidate corrections out of the repository.
It reads the generated dictionary and its canonical inputs, then creates a
GitHub Issue through the authenticated local ``gh`` process only in response to
the desk's explicit Create Issue request.
"""
import argparse
import html
import ipaddress
import json
import mimetypes
from pathlib import Path
import re
import secrets
import subprocess
import sys
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import parse_qs, unquote, urlsplit


TOOL_DIR = Path(__file__).resolve().parent
REPO_ROOT = TOOL_DIR.parents[1]
DICTIONARY_ROOT = REPO_ROOT / 'dictionary'
DESK_JS = TOOL_DIR / 'repair_desk.js'

sys.path.insert(0, str(TOOL_DIR / 'pipeline'))
from anderson import convert as _render_ascii_piece  # noqa: E402


# Review suggestions are local affordances, not accepted corrections. The
# GitHub Issue remains the sole candidate queue; accepted changes still need a
# separate ledger-backed change and tests.
SUGGESTIONS = {
    ('t055-the-trickster-person-who-made-the-country', 622): {
        'before': 'ha:<:: k!&s:<le n@x;-he<mq!etc.',
        'after': 'ha:<:: k!&e:<le n@x;-he<mq!etc.',
        'summary': 'Correct k!&s:<le to k!&e:<le.',
    },
}


def render_miluk(value):
    """Render independently delimited source pieces without crossing spaces.

    The recovered converter searches backward when applying postfix marks. A
    space must therefore be a hard boundary so a stacked mark such as ``k!&``
    cannot attach to a letter in the preceding word.
    """
    return re.sub(r'\S+', lambda match: _render_ascii_piece(match.group(0)), value)


def load_json(path):
    return json.loads(path.read_text(encoding='utf-8'))


def repository_slug(repo_root=REPO_ROOT):
    result = subprocess.run(
        ['git', 'config', '--get', 'remote.origin.url'], cwd=repo_root,
        text=True, capture_output=True, check=False)
    if result.returncode:
        raise RuntimeError('cannot determine the GitHub repository from remote.origin.url')
    remote = result.stdout.strip()
    match = re.search(r'(?:github\.com[/:])([^/]+/[^/]+?)(?:\.git)?$', remote)
    if not match:
        raise RuntimeError('remote.origin.url is not a GitHub repository')
    return match.group(1)


class RepairData:
    """Read-only view of corpus and dictionary context used by the desk."""

    def __init__(self, data_dir=DICTIONARY_ROOT / 'data'):
        corpus = load_json(Path(data_dir) / 'corpus.json')
        dictionary = load_json(Path(data_dir) / 'dictionary.json')
        self.stories = {story['story_id']: story for story in corpus['stories']}
        self.entries = {entry['entry_id']: entry for entry in dictionary['entries']}

    def context(self, story_id, line_number, entry_id=None, form_index=None):
        story = self.stories.get(story_id)
        if story is None:
            raise ValueError('unknown story')
        try:
            line_number = int(line_number)
        except (TypeError, ValueError):
            raise ValueError('line must be an integer') from None
        line = next((item for item in story['lines'] if item['line'] == line_number), None)
        if line is None:
            raise ValueError('unknown story line')

        selected_form = None
        selected_entry = None
        if entry_id:
            selected_entry = self.entries.get(entry_id)
            if selected_entry is None:
                raise ValueError('unknown dictionary entry')
            if form_index is not None:
                try:
                    form_index = int(form_index)
                    selected_form = selected_entry['forms'][form_index]
                except (TypeError, ValueError, IndexError):
                    raise ValueError('unknown attested form') from None

        related_ids = list(line.get('entries', []))
        if selected_entry and entry_id not in related_ids:
            related_ids.insert(0, entry_id)
        related = []
        for related_id in related_ids:
            entry = self.entries.get(related_id)
            if entry is None:
                continue
            item = {
                'entry_id': entry['entry_id'],
                'headword_ascii': entry.get('headword_ascii') or '',
                'headword': entry.get('headword') or '',
                'gloss': entry.get('gloss') or '',
            }
            if related_id == entry_id and selected_form is not None:
                item['selected_form'] = {
                    'index': form_index,
                    'ascii': selected_form.get('ascii') or '',
                    'form': selected_form.get('form') or '',
                    'evidence': selected_form.get('evidence') or '',
                }
            related.append(item)

        current_ascii = line['miluk_ascii']
        suggestion = SUGGESTIONS.get((story_id, line_number), {})
        proposed_ascii = (suggestion.get('after')
                          if suggestion.get('before') == current_ascii
                          else current_ascii)
        return {
            'schema': 'miluk-dictionary-repair-context/1',
            'story': {
                'story_id': story_id,
                'title': story['title'],
                'line': line_number,
            },
            'current': {
                'miluk_ascii': current_ascii,
                'miluk': line['miluk'],
                'english': line.get('english') or '',
            },
            'related_dictionary': related,
            'proposed': {
                'miluk_ascii': proposed_ascii,
                'miluk': render_miluk(proposed_ascii),
                'summary': suggestion.get('summary') or '',
            },
        }


def issue_markdown(context, proposed_ascii, proposed_miluk, summary='', notes=''):
    story = context['story']
    current = context['current']
    related = context['related_dictionary']
    structured = {
        'schema': 'miluk-dictionary-correction-candidate/1',
        'status': 'candidate',
        'target': {
            'source': 'corpus',
            'story_id': story['story_id'],
            'line': story['line'],
            'field': 'miluk_ascii',
        },
        'before': {
            'miluk_ascii': current['miluk_ascii'],
            'miluk': current['miluk'],
        },
        'after': {
            'miluk_ascii': proposed_ascii,
            'miluk': proposed_miluk,
        },
        'related_dictionary': related,
        'reviewer_summary': summary.strip(),
        'reviewer_notes': notes.strip(),
    }
    related_lines = []
    for entry in related:
        line = ('- `{entry_id}` — `{headword_ascii}` / {headword}'
                .format(**entry))
        if entry.get('gloss'):
            line += ' — ' + entry['gloss']
        if entry.get('selected_form'):
            form = entry['selected_form']
            line += ('\n  - selected attested form: `{}` / {} ({})'
                     .format(form['ascii'], form['form'], form['evidence']))
        related_lines.append(line)
    if not related_lines:
        related_lines.append('- No related dictionary entry is currently linked.')

    return """## Dictionary correction candidate

This issue is a review candidate only. Acceptance requires a separate,
ledger-backed and tested correction; creating this issue does not change the
corpus, dictionary, correction ledger, or published site.

| Source | Value |
| --- | --- |
| Story | `{story_id}` — {title} |
| Line | {line_number} |
| English | {english} |

### Proposed correction

{summary}

| | 1990 ASCII | Rendered Miluk |
| --- | --- | --- |
| Before | `{before_ascii}` | {before_miluk} |
| After | `{after_ascii}` | {after_miluk} |

### Related dictionary context

{related}

### Reviewer notes

{notes}

### Structured candidate

```json
{structured}
```
""".format(
        story_id=story['story_id'], title=story['title'],
        line_number=story['line'], english=current['english'] or '—',
        summary=summary.strip() or 'No summary supplied.',
        before_ascii=current['miluk_ascii'], before_miluk=current['miluk'],
        after_ascii=proposed_ascii, after_miluk=proposed_miluk,
        related='\n'.join(related_lines), notes=notes.strip() or 'None.',
        structured=json.dumps(structured, ensure_ascii=False, indent=2),
    )


def create_issue(context, proposed_ascii, repo, summary='', notes='', runner=subprocess.run):
    proposed_ascii = proposed_ascii.strip()
    if not proposed_ascii:
        raise ValueError('proposed miluk_ascii is required')
    if len(proposed_ascii) > 4000 or len(summary) > 300 or len(notes) > 8000:
        raise ValueError('candidate text is too long')
    if proposed_ascii == context['current']['miluk_ascii']:
        raise ValueError('the proposal must differ from the current miluk_ascii')
    proposed_miluk = render_miluk(proposed_ascii)
    body = issue_markdown(context, proposed_ascii, proposed_miluk, summary, notes)
    story = context['story']
    compact = summary.strip() or '{} → {}'.format(
        context['current']['miluk_ascii'], proposed_ascii)
    compact = ' '.join(compact.split())[:120]
    title = '[Dictionary correction] {} line {}: {}'.format(
        story['story_id'].split('-', 1)[0], story['line'], compact)
    command = [
        'gh', 'issue', 'create', '--repo', repo,
        '--title', title, '--body-file', '-',
    ]
    result = runner(command, cwd=REPO_ROOT, input=body, text=True,
                    capture_output=True, check=False)
    if result.returncode:
        detail = (result.stderr or result.stdout or 'gh issue create failed').strip()
        raise RuntimeError(detail)
    match = re.search(r'https://github\.com/[^\s]+/issues/\d+', result.stdout or '')
    if not match:
        raise RuntimeError('gh did not return a GitHub Issue URL')
    return match.group(0), body


def is_loopback_name(value):
    if value == 'localhost':
        return True
    try:
        return ipaddress.ip_address(value).is_loopback
    except ValueError:
        return False


class RepairDeskServer(ThreadingHTTPServer):
    daemon_threads = True

    def __init__(self, address, handler, data, token, repo, issue_runner=subprocess.run):
        super().__init__(address, handler)
        self.repair_data = data
        self.repair_token = token
        self.github_repo = repo
        self.issue_runner = issue_runner


class RepairDeskHandler(BaseHTTPRequestHandler):
    server_version = 'MilukRepairDesk/1'

    def log_message(self, format_string, *args):
        sys.stderr.write('[repair desk] ' + (format_string % args) + '\n')

    def _host_allowed(self):
        host = urlsplit('//' + (self.headers.get('Host') or '')).hostname
        return bool(host and is_loopback_name(host))

    def _origin_allowed(self):
        origin = self.headers.get('Origin')
        if not origin:
            return False
        parsed = urlsplit(origin)
        return (parsed.scheme == 'http' and parsed.hostname and
                is_loopback_name(parsed.hostname) and
                parsed.port == self.server.server_port)

    def _headers(self, status, content_type, length=0):
        self.send_response(status)
        self.send_header('Content-Type', content_type)
        self.send_header('Content-Length', str(length))
        self.send_header('Cache-Control', 'no-store')
        self.send_header('X-Content-Type-Options', 'nosniff')
        self.send_header('Referrer-Policy', 'same-origin')
        self.end_headers()

    def _bytes(self, payload, content_type, status=HTTPStatus.OK):
        self._headers(status, content_type, len(payload))
        self.wfile.write(payload)

    def _json(self, value, status=HTTPStatus.OK):
        payload = (json.dumps(value, ensure_ascii=False) + '\n').encode('utf-8')
        self._bytes(payload, 'application/json; charset=utf-8', status)

    def _error(self, status, message):
        self._json({'error': message}, status)

    def _query_context(self, query):
        values = parse_qs(query, keep_blank_values=True)
        return self.server.repair_data.context(
            values.get('story_id', [''])[0], values.get('line', [''])[0],
            values.get('entry_id', [None])[0], values.get('form_index', [None])[0])

    def _read_json(self):
        try:
            length = int(self.headers.get('Content-Length', '0'))
        except ValueError:
            raise ValueError('invalid request length') from None
        if not 0 < length <= 16384:
            raise ValueError('invalid request length')
        try:
            return json.loads(self.rfile.read(length).decode('utf-8'))
        except (UnicodeDecodeError, json.JSONDecodeError):
            raise ValueError('invalid JSON request') from None

    def do_GET(self):
        if not self._host_allowed():
            self._error(HTTPStatus.FORBIDDEN, 'repair desk accepts loopback hosts only')
            return
        parsed = urlsplit(self.path)
        if parsed.path == '/':
            self.send_response(HTTPStatus.SEE_OTHER)
            self.send_header('Location', '/dictionary/')
            self.send_header('Content-Length', '0')
            self.end_headers()
            return
        if parsed.path == '/__repair/context':
            try:
                self._json(self._query_context(parsed.query))
            except ValueError as error:
                self._error(HTTPStatus.BAD_REQUEST, str(error))
            return
        if parsed.path == '/__repair/desk.js':
            config = 'window.DICTIONARY_REPAIR_DESK_CONFIG = {};\n'.format(
                json.dumps({'token': self.server.repair_token}, ensure_ascii=False))
            payload = config.encode('utf-8') + DESK_JS.read_bytes()
            self._bytes(payload, 'text/javascript; charset=utf-8')
            return
        self._serve_dictionary_file(parsed.path)

    def _serve_dictionary_file(self, request_path):
        relative = unquote(request_path).lstrip('/')
        if relative == 'dictionary':
            self.send_response(HTTPStatus.MOVED_PERMANENTLY)
            self.send_header('Location', '/dictionary/')
            self.send_header('Content-Length', '0')
            self.end_headers()
            return
        if relative.endswith('/'):
            relative += 'index.html'
        candidate = (REPO_ROOT / relative).resolve()
        try:
            candidate.relative_to(DICTIONARY_ROOT.resolve())
        except ValueError:
            self._error(HTTPStatus.NOT_FOUND, 'not found')
            return
        if not candidate.is_file():
            self._error(HTTPStatus.NOT_FOUND, 'not found')
            return
        payload = candidate.read_bytes()
        content_type = mimetypes.guess_type(candidate.name)[0] or 'application/octet-stream'
        if candidate.suffix == '.html' and (
                candidate.parent == DICTIONARY_ROOT / 'stories' or
                candidate.parent == DICTIONARY_ROOT / 'words'):
            marker = b'</body>'
            injection = b'<script src="/__repair/desk.js"></script>\n'
            payload = payload.replace(marker, injection + marker, 1)
        if content_type.startswith('text/') or content_type in {'application/javascript', 'application/json'}:
            content_type += '; charset=utf-8'
        self._bytes(payload, content_type)

    def do_POST(self):
        if not self._host_allowed() or not self._origin_allowed():
            self._error(HTTPStatus.FORBIDDEN, 'same-origin loopback request required')
            return
        if not secrets.compare_digest(
                self.headers.get('X-Repair-Desk-Token') or '',
                self.server.repair_token):
            self._error(HTTPStatus.FORBIDDEN, 'invalid repair desk session')
            return
        parsed = urlsplit(self.path)
        try:
            request = self._read_json()
            context = self.server.repair_data.context(
                request.get('story_id'), request.get('line'),
                request.get('entry_id'), request.get('form_index'))
            proposed_ascii = str(request.get('proposed_ascii') or '').strip()
            if parsed.path == '/__repair/preview':
                if len(proposed_ascii) > 4000:
                    raise ValueError('candidate text is too long')
                self._json({'miluk_ascii': proposed_ascii,
                            'miluk': render_miluk(proposed_ascii)})
                return
            if parsed.path == '/__repair/issues':
                issue_url, _ = create_issue(
                    context, proposed_ascii, self.server.github_repo,
                    str(request.get('summary') or ''), str(request.get('notes') or ''),
                    runner=self.server.issue_runner)
                self._json({'issue_url': issue_url}, HTTPStatus.CREATED)
                return
            self._error(HTTPStatus.NOT_FOUND, 'not found')
        except ValueError as error:
            self._error(HTTPStatus.BAD_REQUEST, str(error))
        except RuntimeError as error:
            self._error(HTTPStatus.BAD_GATEWAY, str(error))


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--port', type=int, default=8000,
                        help='loopback port (default: 8000)')
    parser.add_argument('--repo', default=None,
                        help='GitHub owner/repository (defaults to origin)')
    args = parser.parse_args(argv)
    if not 0 <= args.port <= 65535:
        parser.error('--port must be between 0 and 65535')
    data = RepairData()
    repo = args.repo or repository_slug()
    token = secrets.token_urlsafe(32)
    server = RepairDeskServer(
        ('127.0.0.1', args.port), RepairDeskHandler, data, token, repo)
    url = 'http://127.0.0.1:{}/dictionary/'.format(server.server_port)
    print('Dictionary Repair Desk: ' + url, flush=True)
    print('Candidates become GitHub Issues only after Create GitHub Issue is pressed.',
          flush=True)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()


if __name__ == '__main__':
    main()

File: tools/dictionary/repair_desk.js
/* Local-only Dictionary Repair Desk. Served dynamically by repair_desk.py. */
(function () {
  'use strict';

  var cfg = window.DICTIONARY_REPAIR_DESK_CONFIG || {};
  if (!cfg.token || ['127.0.0.1', 'localhost', '::1'].indexOf(location.hostname) < 0) return;

  function esc(value) {
    return String(value == null ? '' : value).replace(/[&<>"']/g, function (c) {
      return {'&':'&amp;', '<':'&lt;', '>':'&gt;', '"':'&quot;', "'":'&#39;'}[c];
    });
  }

  function changedMarkup(before, after, tag) {
    if (before === after) return esc(before);
    var start = 0;
    while (start < before.length && start < after.length && before[start] === after[start]) start++;
    var endBefore = before.length;
    var endAfter = after.length;
    while (endBefore > start && endAfter > start &&
           before[endBefore - 1] === after[endAfter - 1]) {
      endBefore--; endAfter--;
    }
    return esc(before.slice(0, start)) + '<' + tag + '>' +
      esc((tag === 'del' ? before.slice(start, endBefore) : after.slice(start, endAfter)) || '∅') +
      '</' + tag + '>' + esc(before.slice(endBefore));
  }

  function request(path, options) {
    options = options || {};
    options.headers = Object.assign({'X-Repair-Desk-Token': cfg.token}, options.headers || {});
    return fetch(path, options).then(function (response) {
      return response.json().then(function (payload) {
        if (!response.ok) throw new Error(payload.error || ('Request failed: ' + response.status));
        return payload;
      });
    });
  }

  var style = document.createElement('style');
  style.textContent = [
    '.repair-queue{font:600 .7rem system-ui,sans-serif;color:#e8d4a8;background:#211a0d;border:1px solid #6c511d;border-radius:.35rem;padding:.25rem .5rem;cursor:pointer}',
    '.repair-queue:hover,.repair-queue:focus-visible{color:#0a0a0a;background:#c9952a;outline:none}',
    '.line .repair-queue{grid-column:2;justify-self:start;margin:.25rem 0 .1rem}',
    '.formlist li .repair-queue{margin:.25rem .3rem .25rem 0;white-space:nowrap}',
    '.repair-dialog{width:min(900px,calc(100vw - 2rem));max-height:calc(100vh - 2rem);overflow:auto;color:#f2efe9;background:#101010;border:1px solid #6c511d;border-radius:.7rem;padding:0;box-shadow:0 1.5rem 5rem #000}',
    '.repair-dialog::backdrop{background:rgba(0,0,0,.75)}',
    '.repair-head,.repair-body{padding:1rem 1.2rem}.repair-head{display:flex;justify-content:space-between;gap:1rem;border-bottom:1px solid #262626}',
    '.repair-head h2{margin:0;font-size:.85rem}.repair-close{background:none;color:#9a948a;border:0;font-size:1.5rem;cursor:pointer}',
    '.repair-meta{font: .8rem system-ui,sans-serif;color:#9a948a}.repair-related{font:.78rem system-ui,sans-serif;color:#9a948a;margin:.45rem 0 1rem}',
    '.repair-diff{display:grid;grid-template-columns:1fr 1fr;gap:.8rem;margin:.8rem 0}.repair-card{background:#171717;border:1px solid #262626;border-radius:.45rem;padding:.7rem;min-width:0}',
    '.repair-card h3{font:.7rem system-ui,sans-serif;text-transform:uppercase;letter-spacing:.1em;color:#c9952a;margin:0 0 .35rem}',
    '.repair-card code,.repair-card .mk{display:block;overflow-wrap:anywhere;white-space:pre-wrap}.repair-card del{background:#51262b;color:#ffd7db}.repair-card ins{background:#23412a;color:#d8ffe0;text-decoration:none}',
    '.repair-field{display:block;font:.75rem system-ui,sans-serif;color:#9a948a;margin:.75rem 0 .2rem}.repair-dialog textarea,.repair-dialog input{width:100%;color:#f2efe9;background:#141414;border:1px solid #3a3a3a;border-radius:.4rem;padding:.55rem;font:inherit}',
    '.repair-dialog textarea{min-height:4.5rem;resize:vertical}.repair-actions{display:flex;align-items:center;gap:.7rem;margin-top:1rem;flex-wrap:wrap}',
    '.repair-create{font:600 .8rem system-ui,sans-serif;color:#0a0a0a;background:#c9952a;border:0;border-radius:.4rem;padding:.55rem .8rem;cursor:pointer}.repair-create:disabled{opacity:.45;cursor:not-allowed}',
    '.repair-status{font:.78rem system-ui,sans-serif;color:#9a948a}.repair-status.error{color:#e0a8a8}.repair-status a{color:#e8d4a8}',
    '@media(max-width:650px){.repair-diff{grid-template-columns:1fr}}'
  ].join('');
  document.head.appendChild(style);

  var dialog = document.createElement('dialog');
  dialog.className = 'repair-dialog';
  dialog.innerHTML = '<div class="repair-head"><h2>Dictionary Repair Desk</h2>' +
    '<button class="repair-close" type="button" aria-label="Close">×</button></div>' +
    '<div class="repair-body"><p class="repair-meta"></p><p class="repair-related"></p>' +
    '<div class="repair-diff"><section class="repair-card repair-before"><h3>Before — current source</h3>' +
    '<code></code><span class="mk"></span></section><section class="repair-card repair-after">' +
    '<h3>After — proposed correction</h3><code></code><span class="mk"></span></section></div>' +
    '<label class="repair-field" for="repair-ascii">Proposed miluk_ascii</label>' +
    '<textarea id="repair-ascii" spellcheck="false"></textarea>' +
    '<label class="repair-field" for="repair-summary">Summary</label>' +
    '<input id="repair-summary" type="text" maxlength="300">' +
    '<label class="repair-field" for="repair-notes">Reviewer notes (optional)</label>' +
    '<textarea id="repair-notes" maxlength="8000"></textarea>' +
    '<div class="repair-actions"><button type="button" class="repair-create">Create GitHub Issue</button>' +
    '<span class="repair-status" role="status" aria-live="polite"></span></div></div>';
  document.body.appendChild(dialog);

  var currentContext = null;
  var currentRequest = null;
  var currentRendered = '';
  var previewSequence = 0;
  var asciiInput = dialog.querySelector('#repair-ascii');
  var summaryInput = dialog.querySelector('#repair-summary');
  var notesInput = dialog.querySelector('#repair-notes');
  var createButton = dialog.querySelector('.repair-create');
  var status = dialog.querySelector('.repair-status');

  dialog.querySelector('.repair-close').addEventListener('click', function () { dialog.close(); });

  function updateDiff() {
    if (!currentContext) return;
    var beforeAscii = currentContext.current.miluk_ascii;
    var afterAscii = asciiInput.value;
    var beforeMiluk = currentContext.current.miluk;
    var afterMiluk = currentRendered;
    dialog.querySelector('.repair-before code').innerHTML = changedMarkup(beforeAscii, afterAscii, 'del');
    dialog.querySelector('.repair-after code').innerHTML = changedMarkup(beforeAscii, afterAscii, 'ins');
    dialog.querySelector('.repair-before .mk').innerHTML = changedMarkup(beforeMiluk, afterMiluk, 'del');
    dialog.querySelector('.repair-after .mk').innerHTML = changedMarkup(beforeMiluk, afterMiluk, 'ins');
    createButton.disabled = !afterAscii.trim() || afterAscii.trim() === beforeAscii;
  }

  function preview() {
    var sequence = ++previewSequence;
    return request('/__repair/preview', {
      method: 'POST', headers: {'Content-Type':'application/json'},
      body: JSON.stringify(Object.assign({}, currentRequest, {proposed_ascii: asciiInput.value}))
    }).then(function (payload) {
      if (sequence !== previewSequence) return;
      currentRendered = payload.miluk;
      updateDiff();
    }).catch(function (error) {
      status.textContent = error.message; status.className = 'repair-status error';
    });
  }

  var previewTimer = null;
  asciiInput.addEventListener('input', function () {
    clearTimeout(previewTimer);
    previewTimer = setTimeout(preview, 180);
  });

  function openDesk(source) {
    previewSequence++;
    currentContext = null;
    currentRequest = source;
    createButton.disabled = true;
    status.textContent = 'Loading source context…'; status.className = 'repair-status';
    var query = new URLSearchParams(source).toString();
    request('/__repair/context?' + query).then(function (context) {
      currentContext = context;
      currentRendered = context.proposed.miluk;
      dialog.querySelector('.repair-meta').textContent = context.story.title + ' · ' +
        context.story.story_id + ' · line ' + context.story.line;
      dialog.querySelector('.repair-related').textContent = context.related_dictionary.length ?
        'Related: ' + context.related_dictionary.map(function (entry) {
          var text = entry.entry_id + ' · ' + entry.headword;
          if (entry.selected_form) text += ' · source form ' + entry.selected_form.form;
          return text;
        }).join('; ') : 'No related dictionary entry is currently linked.';
      asciiInput.value = context.proposed.miluk_ascii;
      summaryInput.value = context.proposed.summary;
      notesInput.value = '';
      status.textContent = 'Review the before/after diff. No issue exists yet.';
      status.className = 'repair-status';
      updateDiff();
      if (!dialog.open) dialog.showModal();
    }).catch(function (error) {
      status.textContent = error.message; status.className = 'repair-status error';
      if (!dialog.open) dialog.showModal();
    });
  }

  createButton.addEventListener('click', function () {
    if (!currentContext || createButton.disabled) return;
    createButton.disabled = true;
    status.textContent = 'Creating GitHub Issue through the local gh CLI…';
    status.className = 'repair-status';
    request('/__repair/issues', {
      method: 'POST', headers: {'Content-Type':'application/json'},
      body: JSON.stringify(Object.assign({}, currentRequest, {
        proposed_ascii: asciiInput.value,
        summary: summaryInput.value,
        notes: notesInput.value
      }))
    }).then(function (payload) {
      status.innerHTML = 'Issue created: <a target="_blank" rel="noopener" href="' +
        esc(payload.issue_url) + '">' + esc(payload.issue_url) + '</a>';
    }).catch(function (error) {
      createButton.disabled = false;
      status.textContent = error.message; status.className = 'repair-status error';
    });
  });

  var storyMatch = location.pathname.match(/\/dictionary\/stories\/([^/]+)\.html$/);
  if (storyMatch) {
    document.querySelectorAll('#story .line').forEach(function (line) {
      var number = Number(line.id.slice(1));
      var button = document.createElement('button');
      button.type = 'button'; button.className = 'repair-queue'; button.textContent = 'Queue correction';
      button.addEventListener('click', function () {
        openDesk({story_id: storyMatch[1], line: number});
      });
      line.appendChild(button);
    });
  }

  var entryMatch = location.pathname.match(/\/dictionary\/words\/([^/]+)\.html$/);
  if (entryMatch) {
    var formItems = Array.prototype.slice.call(document.querySelectorAll('.formlist > li'));
    formItems.forEach(function (item, formIndex) {
      var link = item.querySelector('a.formlink');
      if (!link) return;
      var target = new URL(link.href);
      var sourceMatch = target.pathname.match(/\/dictionary\/stories\/([^/]+)\.html$/);
      var lineMatch = target.hash.match(/^#l(\d+)$/);
      if (!sourceMatch || !lineMatch) return;
      var button = document.createElement('button');
      button.type = 'button'; button.className = 'repair-queue'; button.textContent = 'Queue correction';
      button.addEventListener('click', function () {
        openDesk({story_id: sourceMatch[1], line: Number(lineMatch[1]),
                  entry_id: entryMatch[1], form_index: formIndex});
      });
      item.appendChild(button);
    });
  }
})();
